File: src-tauri/helpers/ocr-vision.py
#!/usr/bin/env python3
"""Vision.framework OCR for Goldfish — fast, accurate, GPU-accelerated on Apple Silicon."""
import sys
import Vision
import Quartz
from Foundation import NSURL

if len(sys.argv) < 2:
    print("Usage: ocr-vision.py <image-path>", file=sys.stderr)
    sys.exit(1)

path = sys.argv[1]
url = NSURL.fileURLWithPath_(path)
source = Quartz.CGImageSourceCreateWithURL(url, None)
if source is None:
    print(f"Error: Cannot load image: {path}", file=sys.stderr)
    sys.exit(1)

cgImage = Quartz.CGImageSourceCreateImageAtIndex(source, 0, None)
if cgImage is None:
    print(f"Error: Cannot create CGImage from: {path}", file=sys.stderr)
    sys.exit(1)

request = Vision.VNRecognizeTextRequest.alloc().init()
request.setRecognitionLevel_(1)  # fast
request.setUsesLanguageCorrection_(False)

handler = Vision.VNImageRequestHandler.alloc().initWithCGImage_options_(cgImage, {})
success, error = handler.performRequests_error_([request], None)

if not success:
    print(f"Vision error: {error}", file=sys.stderr)
    sys.exit(1)

for obs in request.results():
    candidates = obs.topCandidates_(1)
    if candidates:
        print(candidates[0].string())

File: src-tauri/helpers/ocr-vision.swift
#!/usr/bin/env swift
// ocr-vision.swift — Fast Vision.framework OCR for Goldfish
// Compile once: swiftc -O -o ocr-vision ocr-vision.swift
// Usage: ./ocr-vision /path/to/image.png

import Cocoa
import Vision

// Get image path from args
guard CommandLine.arguments.count > 1 else {
    fputs("Usage: ocr-vision <image-path>\n", stderr)
    exit(1)
}

let imagePath = CommandLine.arguments[1]

// Load image
guard let image = NSImage(contentsOfFile: imagePath),
      let cgImage = image.cgImage(forProposedRect: nil, context: nil, hints: nil) else {
    fputs("Error: Cannot load image: \(imagePath)\n", stderr)
    exit(1)
}

// Create Vision request
let request = VNRecognizeTextRequest { request, error in
    if let error = error {
        fputs("Vision error: \(error.localizedDescription)\n", stderr)
        exit(1)
    }
    
    guard let observations = request.results as? [VNRecognizedTextObservation] else {
        exit(0)
    }
    
    // Output text line by line
    for observation in observations {
        if let candidate = observation.topCandidates(1).first {
            print(candidate.string)
        }
    }
}

// Use fast mode - good enough for screen text
request.recognitionLevel = .fast
request.usesLanguageCorrection = false
request.recognitionLanguages = ["en-US"]

// Perform OCR
let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
do {
    try handler.perform([request])
} catch {
    fputs("Handler error: \(error.localizedDescription)\n", stderr)
    exit(1)
}
